odoo/addons/base/models/ir_ui_view.py

```diff
@@ -360,13 +360,13 @@ def reset_arch(self, mode='soft'):
         """
         for view in self:
             arch = False
+            # Don't save current arch in previous since we reset, this arch is probably broken
             if mode == 'soft':
                 arch = view.arch_prev
+                view.with_context(no_save_prev=True, lang=None).write({'arch_db': arch})
             elif mode == 'hard' and view.arch_fs:
                 arch = view.with_context(read_arch_from_file=True, lang=None).arch
-            if arch:
-                # Don't save current arch in previous since we reset, this arch is probably broken
-                view.with_context(no_save_prev=True, lang=None).write({'arch_db': arch})
+                view.with_context(no_save_prev=True, lang=None).write({'arch_db': arch, 'arch_updated': False})
```

Comment:
Probably better if you avoid to duplicate the `with_context` code, only the `write` differs